test/e2e/scheduling/hard_constraints.go

```diff
@@ -915,6 +919,285 @@ var _ = SIGDescribe("SchedulingHardConstraints [Serial]", func() {
+			framework.ExpectNoError(e2epod.WaitTimeoutForPodRunningInNamespace(f.ClientSet, antiAffinityPodConf.Name, ns, framework.PollShortTimeout))
+			framework.ExpectEqual(GetPod(f, ns, antiAffinityPodConf.Name).Spec.NodeName, nodeNames[1])
+
+			ginkgo.By("Trying to create failedPod")
+			failedPodConf := pausePodConfig{
+				Name:        "failed-pod-" + string(uuid.NewUUID()),
+				Annotations: WithNMNodeLauncher(WithHardConstraints(GetPodAnnotations(podutil.GuaranteedPod, podutil.Kubelet), []string{interpodaffinity.Name})),
+				Labels:      map[string]string{testLabelKeys[1]: testLabelValues[1]},
+				Resources: &v1.ResourceRequirements{
+					Requests: v1.ResourceList{beardsecond: resource.MustParse("400")},
+					Limits:   v1.ResourceList{beardsecond: resource.MustParse("400")},
+				},
+			}
+			WaitForSchedulerAfterAction(f, createPausePodAction(f, failedPodConf), ns, failedPodConf.Name, false)
+			verifyResult(cs, 3, 1, ns)
+
+			ginkgo.By("Trying to create successpod")
+			successPodConf := pausePodConfig{
+				Name:        "success-pod-" + string(uuid.NewUUID()),
+				Annotations: WithNMNodeLauncher(WithHardConstraints(GetPodAnnotations(podutil.GuaranteedPod, podutil.Kubelet), []string{interpodaffinity.Name})),
+				Labels:      map[string]string{testLabelKeys[0]: testLabelValues[0]},
+				Resources: &v1.ResourceRequirements{
+					Requests: v1.ResourceList{beardsecond: resource.MustParse("400")},
+					Limits:   v1.ResourceList{beardsecond: resource.MustParse("400")},
+				},
+			}
+			framework.ExpectEqual(runPodAndGetNodeName(f, successPodConf), nodeNames[1])
+		})
+	})
+
+	// Test scenario:
+	// 1. Find 2 nodes to run pods, then create nmnodes as same as the 2 nodes and set same extra resource to the 2 nmnodes and 2 nodes.
+	//    Note that nmnode and node on the same host share the same extra resource.
+	//    The scenario looks like this:
+	// ┌────────────┐   ┌────────────┐
+	// │   host 1   │   │   host 2   │
+	// │┌──────────┐│   │┌──────────┐│
+	// ││  node 1  ││   ││  node 2  ││
+	// │└──────────┘│   │└──────────┘│
+	// │┌──────────┐│   │┌──────────┐│
+	// ││ nmnode 1 ││   ││ nmnode 2 ││
+	// │└──────────┘│   │└──────────┘│
+	// │ extra 100% │   │ extra 100% │
+	// └────────────┘   └────────────┘
+	// 2. Create one basicPod[kubelet] with two label and set its node name to the node 1 that needs 40% of the extra resource.
+	// 3. Wait for the pods to be scheduled.
+	// 4. Create one affinityPod[node-manager] with inter-pod affinity to the basicPod[kubelet] by label[0] that needs 40% of the extra resource.
+	// 5. Make sure the affinityPod is scheduled to the nmnode 1 as same as basicPod.
+	// 6. Create one antiAffinityPod[node-manager] with inter-pod anti affinity to the basicPod[kubelet] by label[1] that needs 40% of the extra resource.
+	// 7. Make sure the antiAffinityPod is scheduled to the nmnode 2.
+	// 8. Create one failedPod[kubelet] with label[1] that needs 40% of the extra resource.
```

Comment:
Thoretically, the failedPod (kubelet) will be scheduled to node 1, because node manager and kubelet have independent resource pools and the node 1 still has 60% resources for node-manager. 

However, the troublesome thing is that we don't have a real node manager agent on nodes in the e2e test cluster. So after the scheduler schedules it to node 1, the kubelet may still take over it and reject it due to insufficient resources on the node.